### PR TITLE
Repository sign packages at ingestion

### PR DESCRIPTION
**Pull request by loic-sharma**
Addresses https://github.com/NuGet/Engineering/issues/1328

### CODE REVIEW

File: tests/Validation.PackageSigning.Core.Tests/Support/CertificateIntegrationTestFixture.cs [hunk 1]
```diff
@@ -141,18 +142,35 @@ void CustomizeAsSigningCertificate(X509V3CertificateGenerator generator)
             return IssueCertificate(ca, "Signing", CustomizeAsSigningCertificate).certificate;
         }
 
-        public async Task<UntrustedCertificate> CreateUntrustedSigningCertificateAsync()
+        public async Task<X509Certificate2> CreateUntrustedRootSigningCertificateAsync()
         {
+            var options = IssueCertificateOptions.CreateDefaultForRootCertificateAuthority();
```
Comment (loic-sharma) [inline tests/Validation.PackageSigning.Core.Tests/Support/CertificateIntegrationTestFixture.cs]: /cc @dtivel This is the untrusted root CA used to repository sign packages.

File: tests/Validation.PackageSigning.ProcessSignature.Tests/SignatureValidatorIntegrationTests.cs [hunk 2]
```diff
@@ -614,27 +614,24 @@ public SignatureValidatorIntegrationTests(CertificateIntegrationTestFixture fixt
             Assert.NotNull(result.NupkgUri);
         }
 
-        [Fact(Skip = "Appears to be flaky")]
+        [Fact]
         public async Task StripsRepositorySignatureWithUntrustedSigningCertificate()
         {
             // Arrange
-            var untrustedCertificate = await _fixture.CreateUntrustedSigningCertificateAsync();
+            var certificate = await _fixture.CreateUntrustedRootSigningCertificateAsync();
```
Comment (loic-sharma) [inline tests/Validation.PackageSigning.ProcessSignature.Tests/SignatureValidatorIntegrationTests.cs]: /cc @dtivel This fixes the flaky untrusted test by using a self-signed untrusted certificate, as you suggested.

**Comment by loic-sharma** [comment]
🔔 Ready for review

/cc @dtivel 
/cc @agr

File: src/Validation.PackageSigning.ProcessSignature/SignatureValidator.cs [hunk 3]
```diff
@@ -73,12 +73,31 @@ public class SignatureValidator : ISignatureValidator
...
-                    return await HandleSignedPackageAsync(context);
+                    result = await HandleSignedPackageAsync(context);
+                }
+                else
+                {
+                    result = await HandleUnsignedPackageAsync(context);
+                }
+
+                // Force the validation to fail if the repository signature is expected but missing. The signature
+                // and signing state that are stored in the database may be still valid.
```
Comment (loic-sharma) [inline src/Validation.PackageSigning.ProcessSignature/SignatureValidator.cs]: Please let me know if this comment is unclear as this is a "protocol change". Before, when a repository signature was expected but not found, the validation would fail and the package's signing state would have a `Status` of `PackageSigningStatus.Invalid`. Now, the validation will still fail but the package's `PackageSigningState` may have a `Status` of `PackageSigningStatus.Valid` or `PackageSigningStatus.Unsigned`.

Why was this changed? The `PackageSignatureValidator` verifies that packages have a repository signature, even when repository signing is disabled. This had the side-effect of changing the `PackageSigningState`'s `Status` to `PackageSigningStatus.Invalid`. The `PackageCertificatesValidator` refuses to verify certificates from invalid signatures (the validator fails the validation immediately), which caused it to invalidate all packages. Now that the `PackageSignatureValidator` does not change the `PackageSigningState`'s `Status` when the repository signature is missing, the `PackageCertificatesValidator` will properly verify the package's certificates.

**TLDR**: A signature's validation can fail and have a `PackageSigningState.Status` of `PackageSigningStatus.Valid` or `PackageSigningStatus.Unsigned`.

File: src/NuGet.Services.Validation.Orchestrator/settings.json [hunk 4]
```diff
@@ -80,7 +80,9 @@
         "E2E.SemVer1Stable.*"
       ]
     },
-    "MessageDelay": "00:00:05"
+    "MessageDelay": "00:00:05",
+    "RepositorySigningEnabled": false,
+    "V3ServiceIndexUrl": "https://api.nuget.org/v3/index.json"
```
Comment (dtivel) [inline src/NuGet.Services.Validation.Orchestrator/settings.json]: This should be an environment-specific setting.

File: src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs [hunk 5]
```diff
@@ -2,20 +2,52 @@
...
+
+        public ScanAndSignMessage(
+            OperationRequestType operationRequestType,
+            Guid packageValidationId,
+            Uri blobUri,
+            string v3ServiceIndexUrl,
+            List<string> owners)
+        {
+            if (operationRequestType == OperationRequestType.Scan &&
+                (!string.IsNullOrEmpty(v3ServiceIndexUrl) || (owners != null && owners.Any())))
```
Comment (dtivel) [inline src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs]: `owners != null` seems adequate.

File: src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs [hunk 6]
```diff
@@ -2,20 +2,52 @@
...
             BlobUri = blobUri ?? throw new ArgumentNullException(nameof(blobUri));
+            V3ServiceIndexUrl = v3ServiceIndexUrl ?? throw new ArgumentNullException(nameof(v3ServiceIndexUrl));
+            Owners = owners ?? throw new ArgumentNullException(nameof(owners));
         }
 
         public OperationRequestType OperationRequestType { get; }
         public Guid PackageValidationId { get; }
         public Uri BlobUri { get; }
+        public string V3ServiceIndexUrl { get; }
+        public List<string> Owners { get; }
```
Comment (dtivel) [inline src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs]: public `IReadOnlyList<string>` PackageOwners { get; }

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs [hunk 7]
```diff
@@ -15,120 +15,73 @@
 using NuGetGallery;
 using Xunit;
 
+using Validation.PackageSigning.Helpers;
+using System.Linq;
+using NuGet.Jobs.Validation.Storage;
```
Comment (dtivel) [inline tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Sort with above.

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs [hunk 8]
```diff
@@ -15,120 +15,73 @@
...
+        public async Task DeletesNothingWhenThereIsNoNupkgUrl()
         {
-            var ex = Assert.Throws<ArgumentNullException>(() => new ScanAndSignProcessor(
-                _validatorStateServiceMock.Object,
-                _enqueuerMock.Object,
-                _packageServiceMock.Object,
-                _criteriaEvaluatorMock.Object,
-                _configurationAccessorMock.Object,
-                null));
+            var request = new ValidationRequest(Guid.NewGuid(), 42, "somepackage", "somversion", "https://example.com/package.nupkg");
```
Comment (dtivel) [inline tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: How about `nuget.test` instead of a .com?

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs [hunk 9]
```diff
@@ -283,40 +390,57 @@ public TheStartAsyncMethod()
 
     public class ScanAndSignProcessorFactsBase
     {
+        /*
```
Comment (dtivel) [inline tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Can you remove this commented out code?

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignEnqueuer.cs [hunk 10]
```diff
@@ -40,10 +46,41 @@ public Task EnqueueScanAsync(IValidationRequest request)
...
+            {
+                throw new ArgumentException("The service index URL parameter is required", nameof(v3ServiceIndexUrl));
+            }
+
+            _logger.LogInformation(
+                "Requested scan and sign for package {PackageId} {PackageVersion} using service index {ServiceIndex} and owners {Owners}",
+                request.PackageId,
+                request.PackageVersion,
+                v3ServiceIndexUrl,
+                owners);
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignEnqueuer.cs]: They are not - this will be handled by the downstream job that repository signs the package. This will serialize the owners as a JSON array.

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs (hunk 7 above)
Comment (loic-sharma) [reply tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Woah, not sure what happened there

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs [hunk 11]
```diff
@@ -134,5 +182,31 @@ private bool ShouldSkip(IValidationRequest request, Package package)
 
             return false;
         }
+
+        private async Task<bool> HasRepositorySignature(IValidationRequest request)
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs [hunk 12]
```diff
@@ -101,10 +125,39 @@ public Task CleanUpAsync(IValidationRequest request)
...
+                    return ValidationResult.Succeeded;
+                }
+
+                await _scanAndSignEnqueuer.EnqueueScanAsync(request);
+            }
+            else
+            {
+                var owners = FindPackageOwners(request);
+
+                if (!_configuration.RepositorySigningEnabled)
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs]: I moved things around a little to make this more logical

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs [hunk 13]
```diff
@@ -101,10 +125,39 @@ public Task CleanUpAsync(IValidationRequest request)
...
+                {
+                    _logger.LogInformation(
+                        "Repository signing is disabled. Scanning instead of signing with {ServiceIndex} and {Owners}",
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageSigning/ScanAndSign/ScanAndSignProcessor.cs]: Fixed this. The new logic will log a line for each decision that is made. It should be obvious now whether/why a package is scanned or scanned and signed

File: src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs (hunk 5 above)
Comment (loic-sharma) [reply src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs]: Fixed

File: src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs (hunk 6 above)
Comment (loic-sharma) [reply src/Validation.ScanAndSign.Core/ScanAndSignMessage.cs]: Fixed

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs (hunk 7 above)
Comment (loic-sharma) [reply tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Fixed

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs (hunk 8 above)
Comment (loic-sharma) [reply tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Fixed

File: tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs (hunk 9 above)
Comment (loic-sharma) [reply tests/Validation.PackageSigning.ScanAndSign.Tests/ScanAndSignProcessorFacts.cs]: Fixed
